### PR TITLE
fix(median-blur): preserve float32 precision for kernels 3 and 5

### PR DESCRIPTION
**Pull request by ayasyrev**
#119

## Summary by Sourcery

Update the median_blur router to use native float32 OpenCV paths for kernels 3 and 5 while falling back to quantized uint8 for larger kernels, and extend tests, contracts, docs, benchmarks, and infrastructure accordingly.

New Features:
- Add a dedicated benchmark script for the public median_blur router across uint8/float32 dtypes and kernels 3, 5, and 7.

Bug Fixes:
- Ensure float32 median_blur uses native OpenCV CV_32F for kernel sizes 3 and 5, preserving sub-uint8 precision instead of always quantizing via uint8_io.
- Validate and reject unsupported input dtypes for median_blur, restricting the router to uint8 and float32.

Enhancements:
- Refine the median_blur implementation to route uint8 images through a shared direct/chunked OpenCV uint8 path and float32 images through kernel-aware native or quantized routes.
- Strengthen median_blur tests to cover dtype- and kernel-specific routing, channel edge cases, non-contiguous and read-only inputs, and precision behavior for float32.
- Expand the router contract for median_blur to include both uint8 and float32 dtypes and all supported channel counts with an updated reference description.

Documentation:
- Update README and decorators documentation to reflect the new median_blur routing behavior and uint8_io usage.
- Document the new median_blur benchmark in the benchmarks README.

Tests:
- Replace and extend median_blur unit tests to validate OpenCV parity for uint8 and float32, precision preservation for float32 kernels 3 and 5, quantized fallback behavior for larger kernels, input immutability, and correct error handling for invalid kernels and dtypes.
- Adjust router_contracts tests to reflect median_blur supporting both uint8 and float32 and all channel counts.

Chores:
- Update synthetic router benchmark configuration to treat median_blur as non-uint8-only.

### CODE REVIEW

**Comment by ayasyrev** [comment]
Added explicit `float16` rejection coverage in `8e07009`.

The immutability tests remain separate intentionally: one covers non-contiguous read-only high-channel input and contiguous output; the other covers non-mutation and non-aliasing for ordinary writable input.

The benchmark remains at kernels 3/5/7 because kernel 7 already exercises the `>=7` quantized and high-channel chunked routes. Larger odd kernels use the same branches, and kernel 9 is covered by correctness tests.

File: tests/router_contracts.py [hunk 1]
```diff
@@ -279,14 +276,14 @@ class RouterContract:
     "median_blur": RouterContract(
         name="median_blur",
         kind="function",
-        dtypes=(DTYPE_UINT8,),
+        dtypes=IMAGE_DTYPES,
         layouts=(LAYOUT_HWC,),
-        channels=(CHANNEL_1, CHANNEL_3, CHANNEL_4),
+        channels=IMAGE_CHANNELS,
```
Comment (ternaus) [inline tests/router_contracts.py]: **[P1] Do not advertise the unbounded `>4` channel contract yet**

`IMAGE_CHANNELS` includes the unconstrained `">4"` bucket, while the new tests only exercise 5 channels. On this PR head with OpenCV 5.0.0, both `median_blur(np.zeros((31, 37, 129), np.uint8), 3)` and the equivalent float32 call fail with the `_src0.dims() <= 2` assertion; 128 channels succeeds. Until the separate OpenCV channel-limit follow-up lands, please avoid declaring support for every `>4` input here, or otherwise bound the documented contract to the range actually supported and tested.

File: benchmarks/benchmark_median_blur.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    sizes = ROUTER_HWC_QUICK_HW if args.quick else ROUTER_HWC_FULL_HW
+    rows: list[MedianBlurBenchRow] = []
+    for height, width in sizes:
+        for channels in CHANNELS:
+            shape = (height, width, channels)
+            for dtype in DTYPES:
+                image = _make_image(rng, shape, dtype)
+                for kernel_size in KERNEL_SIZES:
+                    timing = bench_wall_ms(
+                        lambda image=image, kernel_size=kernel_size: albucore.median_blur(image, kernel_size),
```
Comment (ternaus) [inline benchmarks/benchmark_median_blur.py:68]: **[P2] Include a baseline for the regression requirement**

This only records timings for the PR implementation, so it cannot verify issue #119's requirement that the unchanged uint8 and float32 `ksize=7` routes have no confirmed regression greater than 5%. Please either add a base/current comparison mode or attach/check in a before/after report produced from this matrix. Absolute head timings alone cannot establish that acceptance criterion.

**Review by ternaus — COMMENTED** [review]
Two concrete gaps remain in the declared channel contract and the benchmark evidence.

File: tests/router_contracts.py (hunk 1 above)
Comment (ayasyrev) [reply tests/router_contracts.py]: Addressed in c61b52a. The median_blur router contract no longer uses the unbounded IMAGE_CHANNELS bucket and now declares channels 1, 3, and 4 only. The existing 5/9-channel correctness and benchmark cases remain useful extra coverage, but the manifest no longer advertises every >4 input while the OpenCV channel-limit follow-up is pending.

File: benchmarks/benchmark_median_blur.py (hunk 2 above)
Comment (ayasyrev) [reply benchmarks/benchmark_median_blur.py:68]: Addressed in c61b52a by checking in the [0.2.6 vs current report](https://github.com/albumentations-team/albucore/blob/c61b52a2e19c5aa7316758d3c7ae1a466a58d754/benchmarks/results/REPORT_median_blur_0.2.6_vs_current.md). It records the complete 72-cell matrix, environment and baseline-isolation method, plus a 31-sample alternating paired confirmation across all 48 unchanged uint8 and float32 kernel-7 cells. The paired rerun found zero confirmed regressions above 5%; the worst slowdown was 0.35%. The benchmark README now links this evidence.
